### ISSUE TITLE
Handling keyboard popup with ScrollView?

### ISSUE THREAD

**Issue by aksonov** [FaridSafi/react-native-gifted-form#22]
Sometime form could be large and text input could be invisible when keyboard is shown. How to fix it? I saw fixes like here: http://stackoverflow.com/questions/29313244/how-to-auto-slide-the-window-out-from-behind-keyboard-when-textinput-has-focus 

It would be nice to have it as built-in functionality...

**Comment by FaridSafi** [comment]
I implemented such feature https://github.com/FaridSafi/react-native-gifted-form/blob/master/mixins/ContainerMixin.js#L43
If it's not working, then we have a new bug

**Comment by aksonov** [comment]
I guess i should use GiftedForm as container (root) view this way. But my widgets are not direct successors (because of layout, some of them should be grouped inside another view), so they will not be recognized, right? Now I have two GiftedForms with the same name - one for elements inside one view and another for submit button (which is placed at different place).

**Comment by FaridSafi** [comment]
Yes you can use as root. You can put all kinds of views inside GiftedForm, not only widgets

**Comment by aksonov** [comment]
But if i need to place 'input' widgets inside another view and that view will be inside GiftedForm, widgets will not be recognized as widgets (because they are not direct children)...

**Comment by FaridSafi** [comment]
correct, one solution can be to do like https://github.com/FaridSafi/react-native-gifted-form/blob/master/widgets/GroupWidget.js#L20 in your direct children

**Comment by aksonov** [comment]
Thanks, implemented it, but behavior is quite weird, it scrolls to right, not to bottom:
![simulator screen shot 15 2016 19 02 17](https://cloud.githubusercontent.com/assets/1321329/13056726/63e548e6-d417-11e5-9c14-1c1ca7456517.png)

**Comment by FaridSafi** [comment]
weird, what if you replace https://github.com/FaridSafi/react-native-gifted-form/blob/c748128caf005a3aca5a624b1bf5210a92bb3698/mixins/ContainerMixin.js#L51 by 
```js
          this._scrollResponder.scrollWithoutAnimationTo(0, this._pageY + this._y - this._locationY - keyboardHeight + 44);          
```

**Comment by aksonov** [comment]
Works better, thanks, will you commit it? One problem - it doesn't scroll back to top when i focus on upper elements (i.e. restore position)

**Comment by FaridSafi** [comment]
I don't understand why x and y are inverted in your case. Which RN version are you using (I didn't test with last ones) ?

**Comment by aksonov** [comment]
0.19

**Comment by FaridSafi** [comment]
@danielweinmann Are you using 0.19 and experimenting scrollWithoutAnimationTo issue too ?

**Comment by aksonov** [comment]
Also I got two warnings (scrollWithoutAnimationTo looks deprecated now?)
![simulator screen shot 16 2016 12 31 22](https://cloud.githubusercontent.com/assets/1321329/13075137/5c2fa604-d4a9-11e5-9d65-999f807f93ad.png)

**Comment by FaridSafi** [comment]
I didn't test but I guess it can be fixed by scrollTo({x, y, animated: false})

**Comment by FaridSafi** [comment]
Your text input widget with right image looks nice! you should share it :smiley:

**Comment by aksonov** [comment]
I could submit PR for scrollTo and widget enhancement but i can't now, because one PR is waiting already (i don't want to mix everything in one PR)

**Comment by danielweinmann** [comment]
@FaridSafi @aksonov I'm using 0.19 and I do experiment the issue ;)

Actually, I'm using `scrollEnabled={false}` in all my forms right now, since I don't need it so badly and didn't want to take the time to fix it.

**Comment by rdhaliwal** [comment]
@FaridSafi I'm experiencing this issue as well (RN 0.19), and I've found using scrollTo instead of scrollWithoutAnimationTo seems to fix it.

**Comment by aksonov** [comment]
@FaridSafi Could we also restore scroll position once keyboard is disappeared?

**Comment by FaridSafi** [comment]
Thank you guys. I published 0.0.5 with support of react-native 0.20.0 ```ScrollView.scrollTo({x: 0, y: 5, animated: true})```

@aksonov  ```@FaridSafi Could we also restore scroll position once keyboard is disappeared?```
I couldn't manage to do it

**Comment by aksonov** [comment]
@FaridSafi What the problem with it? Have you tried just to subscribe to keyboard events, like described here:
https://github.com/johanneslumpe/react-native-keyboardevents

Current hardcode keyboard height looks not good.

**Comment by FaridSafi** [comment]
@aksonov I can't remember exactly what was the problem with restoring but it might be related to the contentInset of the ScrollView

I didn't try keyboard events when I implemented it. | agree hardcore keyboard height is not good solution.

Sorry I can't work on this right now. If someone has time to improve the scrolling, I recommend to use the ScrollView Props onKeyboardWillShow/onKeyboardWillHide on iOS and onKeyboardDidShow/onKeyboardDidHide on Android (if onKeyboardWillShow doesn't work in Android)

```js
  onKeyboardWillShow(e) {
    console.log((e.endCoordinates ? e.endCoordinates.height : e.end.height));
  },
```
